In [1]:
from pydataset import data

In [2]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)


pandas.core.frame.DataFrame

In [3]:
df

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

Create a column named passing_english that indicates whether each student has a passing grade in english.

In [4]:
df["passing_english"] = df.english >= 70
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

In [5]:
df.sort_values(by='passing_english', ascending=False)

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False

Sort the english grades first by passing_english and then by student name. 

All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically.

 The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)

In [6]:
df.sort_values(by=['passing_english','name'])

name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [7]:
df.sort_values(by=['passing_english','english'])

name  math  english  reading  passing_english
8    Albert    92       62       87            False
11     Alan    92       62       72            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [105]:
df["average_grade"] = round((df.math + df.english + df.reading) / 3).astype(int)
df

name  math  english  reading  passing_english  average_grade
0     Sally    62       85       80             True             76
1      Jane    88       79       67             True             78
2     Suzie    94       74       95             True             88
3     Billy    98       96       88             True             94
4       Ada    77       92       98             True             89
5      John    79       76       93             True             83
6    Thomas    82       64       81            False             76
7     Marie    93       63       90            False             82
8    Albert    92       62       87            False             80
9   Richard    69       80       94             True             81
10    Isaac    92       99       93             True             95
11     Alan    92       62       72            False             75

2 Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [9]:
# data('mpg', show_doc=True) # view the documentation for the dataset
mpg = data('mpg') # load the dataset and store it in a variable

In [10]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [11]:
len(mpg)

234

In [12]:
mpg.size

2574

In [13]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


## Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:
- How many rows and columns are there?
There are 234 rows and 11 columns
- What are the data types of each column?
THE DATA TYPE INFO CAN BE FOUND IN THE GRAPH ABOVE
- Summarize the dataframe with .info and .describe
DONE ABOVE
- Rename the cty column to city.
DONE
- Rename the hwy column to highway.
DONE
- Do any cars have better city mileage than highway mileage?
NO
- Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.
DONE
- Which car (or cars) has the highest mileage difference?
DONE
- Which compact class car has the lowest highway mileage? The best?

- Create a column named average_mileage that is the mean of the city and highway mileage.
- Which dodge car has the best average mileage? The worst?

In [14]:
new_name = {
    "cty": "city",
    "hwy": "highway"
}
new_name

{'cty': 'city', 'hwy': 'highway'}

In [15]:
mpg = mpg.rename(columns=new_name)

In [16]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [17]:
mpg[(mpg.city > mpg.highway)]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

In [18]:
mpg["mileage_difference"] = mpg.highway - mpg.city
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

In [19]:
mpg["mileage_difference"]

1      11
2       8
3      11
4       9
5      10
       ..
230     9
231     8
232    10
233     8
234     9
Name: mileage_difference, Length: 234, dtype: int64

In [106]:
mpg.sort_values(by="mileage_difference", ascending=False).head()

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   
1           audi          a4    1.8  1999    4  auto(l5)   f    18       29   
229   volkswagen      passat    1.8  1999    4  auto(l5)   f    18       29   
36     chevrolet      malibu    3.5  2008    6  auto(l4)   f    18       29   

    fl       class  mileage_difference  average_mileage  
107  c  subcompact                  12             30.0  
223  d  subcompact                  12             35.0  
1    p     compact                  11             23.5  
229  p     midsize                  11             23.5  
36   r     midsize                  11             23.5

In [21]:
# mpg.sort_values(by=["highway"])#.mpg.class.str.startswith("c")

In [107]:
mpg[mpg["class"] == "compact"].sort_values(by="highway", ascending=False).head()

manufacturer    model  displ  year  cyl       trans drv  city  highway fl  \
213   volkswagen    jetta    1.9  1999    4  manual(m5)   f    33       44  d   
197       toyota  corolla    1.8  2008    4  manual(m5)   f    28       37  r   
196       toyota  corolla    1.8  1999    4  manual(m5)   f    26       35  r   
198       toyota  corolla    1.8  2008    4    auto(l4)   f    26       35  r   
195       toyota  corolla    1.8  1999    4    auto(l4)   f    24       33  r   

       class  mileage_difference  average_mileage  
213  compact                  11             38.5  
197  compact                   9             32.5  
196  compact                   9             30.5  
198  compact                   9             30.5  
195  compact                   9             28.5

In [23]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  city  highway fl  \
1         audi    a4    1.8  1999    4    auto(l5)   f    18       29  p   
2         audi    a4    1.8  1999    4  manual(m5)   f    21       29  p   
3         audi    a4    2.0  2008    4  manual(m6)   f    20       31  p   
4         audi    a4    2.0  2008    4    auto(av)   f    21       30  p   
5         audi    a4    2.8  1999    6    auto(l5)   f    16       26  p   

     class  mileage_difference  
1  compact                  11  
2  compact                   8  
3  compact                  11  
4  compact                   9  
5  compact                  10

In [24]:
mpg["average_mileage"] = (mpg.highway +  mpg.city) / 2
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

In [25]:
mpg.sort_values(by="average_mileage", ascending=False)

manufacturer                model  displ  year  cyl       trans drv  city  \
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f    35   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f    33   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f    29   
197       toyota              corolla    1.8  2008    4  manual(m5)   f    28   
196       toyota              corolla    1.8  1999    4  manual(m5)   f    26   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   

     highway fl       class  mileage_difference  average_mileage  
222       44  d  subcompact                   9             39.5  
213       44  d     compact                  11             38.5  
223       41  d  subcompact                  12             35.0  
197       37  r     compact                   9             32.5  
196       35  r     compact                   9             30.5  
..       ... ..         ...                 ...              ...  
127       12  e         suv                   3             10.5  
60        12  e         suv                   3             10.5  
66        12  e      pickup                   3             10.5  
70        12  e      pickup                   3             10.5  
55        12  e      pickup                   3             10.5  

[234 rows x 13 columns]

In [26]:
mpg[mpg.manufacturer.str.startswith("dodge")].sort_values(by="average_mileage",ascending=False)

manufacturer                model  displ  year  cyl       trans drv  city  \
38        dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   
42        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
43        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
39        dodge          caravan 2wd    3.0  1999    6    auto(l4)   f    17   
47        dodge          caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48        dodge          caravan 2wd    4.0  2008    6    auto(l6)   f    16   
40        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
41        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
45        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
46        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
49        dodge    dakota pickup 4wd    3.7  2008    6  manual(m6)   4    15   
53        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
54        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
50        dodge    dakota pickup 4wd    3.7  2008    6    auto(l4)   4    14   
52        dodge    dakota pickup 4wd    3.9  1999    6  manual(m5)   4    14   
63        dodge          durango 4wd    5.7  2008    8    auto(l5)   4    13   
51        dodge    dakota pickup 4wd    3.9  1999    6    auto(l4)   4    13   
73        dodge  ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4    13   
68        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
58        dodge          durango 4wd    3.9  1999    6    auto(l4)   4    13   
59        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
67        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
61        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
69        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
65        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
56        dodge    dakota pickup 4wd    5.2  1999    8  manual(m5)   4    11   
44        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    11   
62        dodge          durango 4wd    5.2  1999    8    auto(l4)   4    11   
72        dodge  ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4    11   
64        dodge          durango 4wd    5.9  1999    8    auto(l4)   4    11   
57        dodge    dakota pickup 4wd    5.2  1999    8    auto(l4)   4    11   
71        dodge  ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4    11   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   

    highway fl    class  mileage_difference  average_mileage  
38       24  r  minivan                   6             21.0  
42       24  r  minivan                   7             20.5  
43       24  r  minivan                   7             20.5  
39       24  r  minivan                   7             20.5  
47       23  r  minivan                   7             19.5  
48       23  r  minivan                   7             19.5  
40       22  r  minivan                   6             19.0  
41       22  r  minivan                   6             19.0  
45       22  r  minivan                   7             18.5  
46       21  r  minivan                   6             18.0  
49       19  r   pickup                   4             17.0  
53       19  r   pickup                   5             16.5  
54       19  r   pickup                   5             16.5  
50       18  r   pickup                   4             16.0  
52       17  r   

# #############################################################

In [27]:
mammals = data('Mammals')

In [28]:
data('Mammals', show_doc=True) # view the documentation for the dataset

Mammals

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Garland(1983) Data on Running Speed of Mammals

### Description

Observations on the maximal running speed of mammal species and their body
mass.

### Usage

    data(Mammals)

### Format

A data frame with 107 observations on the following 4 variables.

weight

Body mass in Kg for "typical adult sizes"

speed

Maximal running speed (fastest sprint velocity on record)

hoppers

logical variable indicating animals that ambulate by hopping, e.g. kangaroos

specials

logical variable indicating special animals with "lifestyles in which speed
does not figure as an important factor": Hippopotamus, raccoon (Procyon),
badger (Meles), coati (Nasua), skunk (Mephitis), man (Homo), porcupine
(Erithizon), oppossum (didelphis), and sloth (Bradypus)

### Details

Used by Chappell (1989) and Koenker, Ng and Portnoy (1994) to illustrate the
fitting of piecewise linear curves.

### Source

Garland, T. (

In [111]:
mammals

weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    130.000   70.0    False     False
25    120.000   80.0    False     False
26    120.000   61.0    False     False
27    110.000   33.0    False     False
28    100.000   64.0    False     False
29     85.000   55.0    False     False
30     80.000   65.0    False     False
31     72.000   56.0    False     False
32     70.000   45.0    False     False
33     65.000   60.0    False     False
34     62.000   81.0    False     False
35     50.000  100.0    False     False
36     50.000   60.0    False     False
37     50.000   40.0    False     False
38     50.000   47.0    False     False
39     37.000  105.0    False     False
40     35.000   80.0    False     False
41     34.000   97.0    False     False
42     30.000   97.0    False     False
43     30.000   80.0    False     False
44     30.000   45.0    False     False
45     20.000   81.0    False     False
46    400.000   40.0    False     False
47    300.000   48.0    False     False
48    230.000   56.0    False     False
49    150.000   59.0    False     False
50    135.000   48.0    False     False
51     65.000   65.0    False     False
52     60.000   60.0    False     False
53     55.000  110.0    False     False
54     45.000   50.0    False     False
55     40.000   64.0    False     False
56     25.000   67.0    False     False
57     20.000   70.0    False     False
58     16.000   65.0    False     False
59     12.000   24.0    False      True
60     11.000   30.0    False      True
61     10.000   56.0    False     False
62      7.000   60.0    False     False
63      6.000   72.0    False     False
64      5.000   64.0    False     False
65      5.000   27.0    False      True
66      3.000   16.0    False      True
67    127.000   32.0    False     False
68     70.000   40.0    False      True
69     13.000   37.0    False      True
70      9.000    3.2    False      True
71      4.000   16.0    False     False
72      0.600   36.0    False     False
73      0.600   20.0    False     False
74      0.550   27.0    False     False
75      0.500   18.0    False     False
76      0.400   20.0    False     False
77      0.300   13.0    False     False
78      0.250    9.7    False     False
79      0.220   15.0    False     False
80      0.110    9.0    False     False
81      0.100   17.0    False     False
82      0.056   21.0     True     False
83      0.050   11.0    False     False
84      0.045   16.0    False     False
85      0.035   32.0     True     False
86      0.035   14.0     True     False
87      0.030    6.8    False     False
88      0.030    9.1    False     False
89      0.025   11.0    False     False
90      0.025    8.6    False     False
91      0.018    8.9    False     False
92      0.016   13.0    False     False
93      0.100    4.0    False     False
94      0.100    2.4    False     False
95      0.016    3.6    False     False
96      4.600   64.0     True     False
97      4.400   72.0     True     False
98      4.000   72.0     True     False
99      3.500   56.0     True     False
100    

In [115]:
mammals[mammals.speed == mammals.speed.max()]

weight  speed  hoppers  specials
53    55.0  110.0    False     False

In [30]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [101]:
mammals.shape[0]

107

In [32]:
mammals.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

In [33]:
mammals.describe()

weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

In [39]:
mammals.sort_values(by="speed", ascending=False).head(5)

weight  speed  hoppers  specials
53    55.0  110.0    False     False
39    37.0  105.0    False     False
35    50.0  100.0    False     False
41    34.0   97.0    False     False
42    30.0   97.0    False     False

In [68]:
mammals.specials.sum()

10

In [60]:
mammals[mammals.speed > mammals.speed.median()].head()

weight  speed  hoppers  specials
5    400.0   70.0    False     False
6    350.0   70.0    False     False
7    300.0   64.0    False     False
8    260.0   70.0    False     False
11  1000.0   60.0    False     False

the above code is rendering an error

In [53]:
mammals.speed.mean()

46.20841121495327

In [80]:
mammals.speed.tail()

103    40.0
104    65.0
105     7.4
106    13.0
107     1.6
Name: speed, dtype: float64

In [65]:
mammals.sort_values(by='hoppers', ascending=False).sum()

weight      29819.635
speed        4944.300
hoppers        11.000
specials       10.000
dtype: float64

### there are 11 hoppers

In [72]:
mammals.hoppers.sum()

11

How many animals are hoppers that are above the median speed? What percentage is this?

In [91]:
mammals[(mammals.speed > mammals.speed.median()) & (mammals.hoppers == True)].hoppers.sum()#.sort_values(by="speed",ascending=False)

7

In [103]:
((mammals[(mammals.speed > mammals.speed.median()) & (mammals.hoppers == True)].hoppers.sum()) / mammals.shape[0]) * 100#mammals.size

6.5420560747663545